### DISCUSSION TITLE
How clear_masks and set_masks functions handle multiple, non-parametric object edge_masks.

### DISCUSSION THREAD

**Discussion by CXX1113** [pyg-team/pytorch_geometric#8069] [Q&A]
I am building a explainer algorithm. After initializing an edge_mask parameter, ie edge_mask = Parameter(torch.randn(5). If I want to perform some operations on it, including: edge_mask2 = edge_mask +10, edge_mask3 = F.dropout(edge_mask), edge_mask4 = F.dropout(edge_mask), then edge_mask , edge_mask2, edge_mask3, edge_mask4 will be used in the forward function of the  explainer.
My issues is as follows:
In PyG's torch_geometric.explain.algorithm.utils. clear_masks, set_masks,
1. I notied that set_masks will convert each mask into a parameter object, that is, edge_mask2, edge_mask3, and edge_mask4 will be converted into parameter objects. In this case, I am worried about whether it will have a negative impact on edge_mask update. Although I know edge_mask2, edge_mask3, edge_mask4 have not been added to the optimizer, I am still worried whether this will lead to wrong gradient calculation.
2. Because I have 4 edge_masks here, If the code is:
```
set_mask(edge_mask, ...)
y_hat = model(x, edge_index)
clear_masks(model)
set_mask(edge_mask2, ...)
y_hat2 = model(x, edge_index)
clear_masks(model)
set_mask(edge_mask3, ...)
y_hat3 = model(x, edge_index)
clear_masks(model)
set_mask(edge_mask4, ...)
y_hat4 = model(x, edge_index)
clear_masks(model)
```
So, after calling clear_masks twice, will the gradient information of edge_mask be calculated correctly?

**Comment by wsad1** [comment]
Sorry I don't fully get the question.
If you want to compute a gradient for the `edge_mask` parameter you would have to call `loss.backward` before you clear it and set another mask.
